### PR TITLE
Add EDIFHierCellInst.isUniquified()

### PR DESCRIPTION
**Pull request by eddieh-xlnx**
Add the ability to determine if an `EDIFHierCellInst` is unique across the netlist, with a view to enabling algorithms (such as #910) to decide whether it is safe to modify.

To answer the `EDIFHierCellInst.isUniquified()` question, we need to know whether all `EDIFCellInst`-s that make up `EDIFHierCellInst` are themselves individually unique (folded, non-hierarchical) instantiations across the netlist.

To answer this `EDIFCellInst.isUniquified()` question, we need to track how many `EDIFCellInst`s exist for each `EDIFCell` which is also implemented in this PR.

Take for example the hierarchical path:
```
top/foo/bar/baz
```
To determine if it's safe to modify the `baz` instantiation (i.e. to change its parameters/LUT masks, or to change its port connections) we must determine that its parent `bar` instantiation is unique, followed by its parent `foo` instantiation being unique (and assume that `top` must be unique).

### CODE REVIEW

File: src/com/xilinx/rapidwright/edif/EDIFCell.java [hunk 1]
```diff
@@ -656,5 +662,33 @@ public boolean equals(Object o) {
...
+
+    /**
+     * Atomically decrement instance count of this cell.
+     */
+    public void decrementInstanceCount() {
+        instanceCountUpdater.getAndDecrement(this);
+    }
+
+    /**
+     * Atomically decrement instance count of this cell.
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFCell.java:681]:
```suggestion
     * @return The number of times this cell has been instantiated.
```

File: test/src/com/xilinx/rapidwright/edif/TestEDIFCell.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Test
+    public void testIsUniquifiedFalse() {
+        Design design = RapidWrightDCP.loadDCP("picoblaze4_ooc_X6Y60_X6Y65_X10Y60_X10Y65.dcp");
+        EDIFNetlist netlist = design.getNetlist();
+        Assertions.assertTrue(netlist.getTopCell().isUniquified());
+        EDIFCell picoblazeTop = netlist.getCell("picoblaze_top");
+        Assertions.assertFalse(picoblazeTop.isUniquified());
+        Assertions.assertEquals(4, picoblazeTop.getInstanceCount());
+
```
Comment (eddieh-xlnx) [inline test/src/com/xilinx/rapidwright/edif/TestEDIFCell.java:61]:
```suggestion
```

**Comment by clavin-xlnx** [comment]
I think there are still some considerations that need to be reasoned about in order to make this feature a reality that we can depend on.  

1. We will need someway to enforce copy constructors on primitive and macro cells when they are loaded/instantiated in a design so we don't end up changing the values inside a static library of cells.  Obviously, the primitives and macros don't need 'unique-ification' but modifying a cell that doesn't belong to the current design would be bad practice and if two or more designs are loaded at the same time, the instance count would be invalid.  
2. [`public EDIFCellInst removeCellInst(String name)` in `EDIFCell`](https://github.com/Xilinx/RapidWright/blob/f4241d89e65f1ed981f005d175fd41de873a7a6c/src/com/xilinx/rapidwright/edif/EDIFCell.java#L293) would need to be updated accordingly anytime an instance is removed.
3. We may need to think about black boxes and make sure they are handled appropriately without breaking some assumptions of this approach.

**Comment by eddieh-xlnx** [comment]
Sorry, I just cherry-picked some updates I forgot to move back into this branch.

Just to make sure we're on the same page: the purpose of this PR is to allow the determination of whether a `EDIFHierCellInst` is the one-and-only unique instantiation of an `EDIFCell`. The way we do this is to count how many times an `EDIFCellInst`-s instantiates a particular `EDIFCell`, and for `EDIFHierCellInst` to check all instances in its path are unique in this manner.

It does not make sense to call this on an `EDIF{Hier,}CellInst` of a primitive (in fact, we could explicitly block this with an assertion, exception, or false). I cannot see a reason to expect anyone to modify any part of the `EDIFCell` interface or contents of such primitives. You would, however, expect to modify the properties of the `EDIFCellInst` -- and for that, you would want to check that the parent(s) of such a cell instance to be fully uniquified.

Take for example:
```
module foo;
LUT1 #(.INIT(...)) bar();
endmodule
```
It does not make sense to query if `LUT1` is unique, but it does make sense to see if `foo` is (hierarchically) unique in order to determine if the `INIT` parameter can be modified.

In response to your points:
> 1. We will need someway to enforce copy constructors on primitive and macro cells when they are loaded/instantiated in a design so we don't end up changing the values inside a static library of cells. Obviously, the primitives and macros don't need 'unique-ification' but modifying a cell that doesn't belong to the current design would be bad practice and if two or more designs are loaded at the same time, the instance count would be invalid.

Yes, that is a valid point, but one I think should be considered out of the scope of this PR. I believe we have a broader issue with primitives from the static/singleton library not being copied into each design, in which case (and when there are multiple netlists present) then the unique-ness or instance count values are meaningless. Perhaps the best way would be for primitives to error out when asked this question (which is of questionable value to begin with)?

> 2. [public EDIFCellInst removeCellInst(String name) in EDIFCell](https://github.com/Xilinx/RapidWright/blob/f4241d89e65f1ed981f005d175fd41de873a7a6c/src/com/xilinx/rapidwright/edif/EDIFCell.java#L293) would need to be updated accordingly anytime an instance is removed.

This is an excellent point. Currently, I'm hooking into `EDIFCellInst.setCellType()` to increment/decrement the instance count. However, that doesn't capture whether (a) that `EDIFCellInst` is actually attached to an `EDIFCell` (or indeed, if that `EDIFCell` is eventually attached to the top cell of a particular netlist) nor (b) the instance count is not decremented when this cell gets garbage collected, for whatever (other) reason.
This needs more thought, and `EDIFCell.{add,remove}CellInst()` could be a solution here.

> 3. We may need to think about black boxes and make sure they are handled appropriately without breaking some assumptions of this approach.

Can you elaborate on your concerns with blackboxes please? Blackbox-es are `EDIFCell`-s just like white-boxes but have a hollow interior. They can still be instantiated as normal, in which case there can still be count of how many times it gets instantiated, and this state of unique-ness can still be used to determine whether an `EDIFCellInst` can be modified. Nor does it make sense to query the uniqueness `EDIFCellInst` present inside a blackbox, since blackboxes cannot have anything inside.

**Comment by eddieh-xlnx** [comment]
> > 3\. [public EDIFCellInst removeCellInst(String name) in EDIFCell](https://github.com/Xilinx/RapidWright/blob/f4241d89e65f1ed981f005d175fd41de873a7a6c/src/com/xilinx/rapidwright/edif/EDIFCell.java#L293) would need to be updated accordingly anytime an instance is removed.
> 
> This is an excellent point. Currently, I'm hooking into `EDIFCellInst.setCellType()` to increment/decrement the instance count. However, that doesn't capture whether (a) that `EDIFCellInst` is actually attached to an `EDIFCell` (or indeed, if that `EDIFCell` is eventually attached to the top cell of a particular netlist) nor (b) the instance count is not decremented when this cell gets garbage collected, for whatever (other) reason. This needs more thought, and `EDIFCell.{add,remove}CellInst()` could be a solution here.

It looks like handling removals is not so straightforward. In the common case (loading a "minimal" EDIF into memory, where all cells present are instantiated at least once) then each `EDIFCell.addCellInst()` can be assumed to increment the instance count of the `EDIFCell` it instantiates.

However, when we start considering `EDIFCell.removeCellInst()` then things get tricky. Detaching an `EDIFCellInst` from its parent cell would mean that the `EDIFCell` being instantiated should have its instance count decremented. But what about the cell type of *its* instances, and so on, which may also be detached now?

And then how can we elegantly handle `EDIFCell.addCellInst()` for things that were previously detached -- should the whole transitive tree of cells being instantiated also have their instance count bumped? Can double-incrementing be efficiently avoided when the common case uses the same routines?

**Comment by eddieh-xlnx** [comment]
Okay, I think I finally have something I think could work.

`EDIFCellInst`-s will now increment the instance count of the `EDIFCell` it instantiates only if:
1. It has a `parentCell` (the cell it belongs to)
2. It has has a `cellType` (the cell it instantiates)

Attaching to both of those causes the instance count to increment, and detaching either causes it to decrement.

`EDIFCell.removeCellInst()` now sets `EDIFCellInst.parentCell` to be null (indicating detachment) which in turn decrements the instance count of `EDIFCell.cellType`. Unlike an early implementation, we do not recursively traverse into `EDIFCell.cellType` to decrement downstream instances.

`EDIFCell.addCellInst()` does the opposite, and because we check for `parentCell` and `cellType` then I don't believe double incrementing occurs.

File: src/com/xilinx/rapidwright/edif/EDIFCell.java [hunk 3]
```diff
@@ -62,6 +63,14 @@ public class EDIFCell extends EDIFPropertyObject {
 
     private EDIFName view = DEFAULT_VIEW;
 
+    /**
+     * An atomically updated variable to track the number of `EDIFCellInst`
+     * objects (attached to a parent cell) that instantiate this cell.
+     */
+    private volatile int instanceCount = 0;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFCell.java:70]: `directInstanceCount`?
`explicitInstanceCount`?
`foldedInstanceCount`?

File: src/com/xilinx/rapidwright/edif/EDIFCell.java [hunk 4]
```diff
@@ -292,8 +301,11 @@ public EDIFCellInst removeCellInst(EDIFCellInst cellInstance) {
 
     public EDIFCellInst removeCellInst(String name) {
         if (instances == null) return null;
-        trackChange(EDIFChangeType.CELL_INST_REMOVE, name);
-        return instances.remove(name);
+        EDIFCellInst removedInstance = instances.remove(name);
+        if (removedInstance != null) {
+            removedInstance.setParentCell(null);
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/edif/EDIFCell.java:306]:
```suggestion
            assert(removedInstance.getParentCell() == this);
            removedInstance.setParentCell(null);
```

File: src/com/xilinx/rapidwright/edif/EDIFCell.java (hunk 3 above)
Comment (eddieh-xlnx) [reply src/com/xilinx/rapidwright/edif/EDIFCell.java]: I've gone with `nonHierInstantiationCount` to make it clear that it's most definitely not `Hier`.

**Pull request merged**